### PR TITLE
feat: add invitation code for signup feature

### PR DESCRIPTION
**Pull request by leo220yuyaodog**
https://github.com/casdoor/casdoor/assets/71440988/1f5fde5e-e99f-4175-90a7-13e496f8f3d3

### CODE REVIEW

File: controllers/application.go [hunk 1]
```diff
@@ -245,3 +245,27 @@ func (c *ApiController) DeleteApplication() {
...
 }
+
+// AddInvitationCode
+// @Title AddInvitationCode
+// @Tag Application API
+// @Description add an invitation code
+// @Param   body   body   string  true    "The details of the application"
+// @Success 200 {object} string The Response object
+// @router /add-invitation-code [post]
+func (c *ApiController) AddInvitationCode() {
```
Comment (hsluoyz) [inline controllers/application.go:256]: No need for new API

File: object/application.go [hunk 2]
```diff
@@ -58,6 +58,7 @@ type Application struct {
 	GrantTypes          []string        `xorm:"varchar(1000)" json:"grantTypes"`
 	OrganizationObj     *Organization   `xorm:"-" json:"organizationObj"`
 	Tags                []string        `xorm:"mediumtext" json:"tags"`
+	InvitationCodes     []string        `xorm:"mediumtext" json:"invitationCodes"`
```
Comment (hsluoyz) [inline object/application.go:61]: `varchar(200)` is enough

File: object/application.go [hunk 3]
```diff
@@ -534,3 +540,16 @@ func applicationChangeTrigger(oldName string, newName string) error {
 
 	return session.Commit()
 }
+
+func AddInvitationCode(application *Application) ([]string, error) {
```
Comment (hsluoyz) [inline object/application.go:544]: Remove it

File: object/check.go [hunk 4]
```diff
@@ -124,6 +124,18 @@ func CheckUserSignup(application *Application, organization *Organization, form
 		}
 	}
 
+	if application.IsSignupItemVisible("Invitation code") {
```
Comment (hsluoyz) [inline object/check.go:127]: logic is not correct. Only need to check `application.InvitationCodes`'s length > 0

File: routers/router.go [hunk 5]
```diff
@@ -165,6 +165,7 @@ func initAPI() {
 	beego.Router("/api/update-application", &controllers.ApiController{}, "POST:UpdateApplication")
 	beego.Router("/api/add-application", &controllers.ApiController{}, "POST:AddApplication")
 	beego.Router("/api/delete-application", &controllers.ApiController{}, "POST:DeleteApplication")
+	beego.Router("/api/add-invitation-code", &controllers.ApiController{}, "POST:AddInvitationCode")
```
Comment (hsluoyz) [inline routers/router.go:168]: Remove

File: util/random.go [hunk 6]
```diff
@@ -23,3 +23,7 @@ func GenerateClientId() string {
 func GenerateClientSecret() string {
 	return randstr.Hex(20)
 }
+
+func GenerateInvitationCode() string {
```
Comment (hsluoyz) [inline util/random.go:27]: Remove it

File: web/src/ApplicationEditPage.js [hunk 7]
```diff
@@ -13,7 +13,21 @@
 // limitations under the License.
 
 import React from "react";
-import {Button, Card, Col, ConfigProvider, Input, Popover, Radio, Result, Row, Select, Switch, Upload} from "antd";
+import {
+  Button,
+  Card,
+  Col,
+  ConfigProvider,
```
Comment (hsluoyz) [inline web/src/ApplicationEditPage.js:20]: Don't change this

File: web/src/ApplicationEditPage.js [hunk 8]
```diff
@@ -813,18 +831,77 @@ class ApplicationEditPage extends React.Component {
...
-                  title={i18next.t("application:Signup items")}
-                  table={this.state.application.signupItems}
-                  onUpdateTable={(value) => {this.updateApplicationField("signupItems", value);}}
-                />
-              </Col>
-            </Row>
+            <React.Fragment>
+              <Row style={{marginTop: "20px"}} >
+                <Col style={{marginTop: "5px"}} span={(Setting.isMobile()) ? 22 : 2}>
+                  {Setting.getLabel(i18next.t("application:Signup items"), i18next.t("application:Signup items - Tooltip"))} :
```
Comment (hsluoyz) [inline web/src/ApplicationEditPage.js:837]: Don't change

File: web/src/locales/ar/data.json [hunk 9]
```diff
@@ -58,6 +58,9 @@
     "Grant types - Tooltip": "Select which grant types are allowed in the OAuth protocol",
     "Incremental": "Incremental",
     "Input": "Input",
+    "Invitation code": "Invitation code",
+    "Invitation code - Tooltip": "Invitation code - Tooltip",
+    "Invitation code copied to clipboard successfully": "Invitation code copied to clipboard successfully",
```
Comment (hsluoyz) [inline web/src/locales/ar/data.json:63]: No need to update i18n

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/src/ApplicationEditPage.js (hunk 8 above)
Comment (leo220yuyaodog) [reply web/src/ApplicationEditPage.js:822]: For showing invitation codes

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**